extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -141,185 +76,217 @@ export abstract class AzureAuth implements vscode.Disposable {
+		const baseTokens = await this.getSavedToken(this.commonTenant, this.metadata.settings.microsoftResource, account.key);
+		if (!baseTokens) {
+			Logger.error('User had no base tokens for the basic resource registered. This should not happen and indicates something went wrong with the authentication cycle');
+			const msg = localize('azure.noBaseToken', 'Something failed with the authentication, or your tokens have been deleted from the system. Please try adding your account to Azure Data Studio again.');
+			account.isStale = true;
+			throw new AzureAuthError(msg, 'No base token found', undefined);
+		}
+		// Let's try to convert the access token type, worst case we'll have to prompt the user to do an interactive authentication.
+		const result = await this.refreshToken(tenant, resource, baseTokens.refreshToken);
+		if (result.accessToken) {
+			return {
+				...result.accessToken,
 				tokenType: 'Bearer'
 			};
 		}
+		return undefined;
+	}
 
-		if (azureAccount.properties.subscriptions) {
-			azureAccount.properties.subscriptions.forEach(subscription => {
-				// Make sure that tenant has information populated.
-				if (response[subscription.tenantId]) {
-					response[subscription.id] = {
-						...response[subscription.tenantId]
-					};
-				}
-			});
+
+
+	protected abstract async login(tenant: Tenant, resource: Resource): Promise<{ response: OAuthTokenResponse, authComplete: Deferred<void> }>;
+
+	/**
+	 * Refreshes a token, if a refreshToken is passed in then we use that. If it is not passed in then we will prompt the user for consent.
+	 * @param tenant
+	 * @param resource
+	 * @param refreshToken
+	 */
+	protected async refreshToken(tenant: Tenant, resource: Resource, refreshToken: RefreshToken | undefined): Promise<OAuthTokenResponse> {
+		if (refreshToken) {
+			const postData: RefreshTokenPostData = {
+				grant_type: 'refresh_token',
+				client_id: this.clientId,
+				refresh_token: refreshToken.token,
+				tenant: tenant.id,
+				resource: resource.endpoint
+			};
+
+			return this.getToken(tenant, resource, postData);
 		}
 
-		return response;
+		return this.handleInteractionRequired(tenant, resource);
 	}
 
-	public async clearCredentials(account: azdata.AccountKey): Promise<void> {
-		try {
-			return this.deleteAccountCache(account);
-		} catch (ex) {
-			const msg = localize('azure.cacheErrrorRemove', "Error when removing your account from the cache.");
-			vscode.window.showErrorMessage(msg);
-			Logger.error('Error when removing tokens.', ex);
+	protected async getToken(tenant: Tenant, resource: Resource, postData: AuthorizationCodePostData | TokenPostData | RefreshTokenPostData): Promise<OAuthTokenResponse> {
+		const tokenUrl = `${this.loginEndpointUrl}${tenant.id}/oauth2/token`;
+		const response = await this.makePostRequest(tokenUrl, postData);
+
+		if (response.data.error === 'interaction_required') {
```

Comment:
response.data is guaranteed to exist? You're checking for it being undefined below so I assume this needs it too.

Reply:
Yeah response.data is returned by axios. I'll cleanup my undefined checks.